### PR TITLE
EssenceShopPrice feature

### PR DESCRIPTION
**Pull request by UpFault**
added option to view how much essence will cost from bazaar in lore of Essence Shop Items

### CODE REVIEW

**Comment by kevinthegreat1** [comment]
Thanks for your great pr’s. As you may see, we currently have quite a few pr’s awaiting review so this might take a while.

**Comment by UpFault** [comment]
> Thanks for your great pr’s. As you may see, we currently have quite a few pr’s awaiting review so this might take a while.

no worries

**Comment by UpFault** [comment]
ZkWnrwz
fC9jrWU
UjRQHSC


Figured I'd drop these here

**Comment by UpFault** [comment]
Also I should probably move the config value over to Item Tooltip in the general category rather than UI & Visuals(don't know what I was thinking here.)

**Comment by Fluboxer** [comment]
May as well take it to your own visitor pr so they have a child - if you lack essence to buy thing, right clicking would just open bz with needed essence in your clipboard

**Comment by kevinthegreat1** [comment]
Hi, these pr's still need rebasing since they contain unrelated commits. Here's how you can rebase the pr's. I'm using #737 as an example. First, find the first commit in your pr and selected it, as shown in the first image. Right click and select `Interactively Rebase from Here`. Then select all commits except the ones that you want to keep and select `Drop`, and you should end up like the second image. Finally click `Start Rebasing`.
Step 1
Step 2

File: src/main/java/de/hysky/skyblocker/skyblock/item/tooltip/adders/EssenceShopPrice.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import net.minecraft.item.ItemStack;
+import net.minecraft.screen.slot.Slot;
+import net.minecraft.text.Text;
+import net.minecraft.util.Formatting;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.text.DecimalFormat;
+import java.text.DecimalFormatSymbols;
```
Comment (kevinthegreat1) [inline src/main/java/de/hysky/skyblocker/skyblock/item/tooltip/adders/EssenceShopPrice.java:19]:
```suggestion
import java.text.NumberFormat;
```

File: src/main/java/de/hysky/skyblocker/skyblock/item/tooltip/adders/EssenceShopPrice.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.List;
+import java.util.Locale;
+import java.util.OptionalLong;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class EssenceShopPrice extends TooltipAdder {
+	private static final Logger LOGGER = LoggerFactory.getLogger(EssenceShopPrice.class);
+	private static final Pattern ESSENCE_PATTERN = Pattern.compile("Cost (?<amount>[\\d,]+) (?<type>[A-Za-z]+) Essence");
+	private static final DecimalFormat DECIMAL_FORMAT = new DecimalFormat("#,###.#", DecimalFormatSymbols.getInstance(Locale.ENGLISH));
```
Comment (kevinthegreat1) [inline src/main/java/de/hysky/skyblocker/skyblock/item/tooltip/adders/EssenceShopPrice.java:29]:
```suggestion
	private static final NumberFormat DECIMAL_FORMAT = NumberFormat.getInstance(Locale.US);
```

File: src/main/java/de/hysky/skyblocker/skyblock/item/tooltip/adders/EssenceShopPrice.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+		for (String essenceType : ESSENCE_TYPES) {
+			JsonElement price = data.get("ESSENCE_" + essenceType);
+			if (price == null || !price.isJsonObject()) continue;
+			JsonElement sellPrice = price.getAsJsonObject().get("sellPrice");
+			if (sellPrice == null) continue;
+			if (sellPrice.isJsonPrimitive()) {
+				ESSENCE_PRICES.put(essenceType, sellPrice.getAsLong());
+			}
+		}
+		LOGGER.info("[Skyblocker] Refreshed essence prices.");
```
Comment (kevinthegreat1) [inline src/main/java/de/hysky/skyblocker/skyblock/item/tooltip/adders/EssenceShopPrice.java:47]:
```suggestion
```

File: src/main/java/de/hysky/skyblocker/config/categories/GeneralCategory.java [hunk 4]
```diff
@@ -222,6 +222,13 @@ public static ConfigCategory create(SkyblockerConfig defaults, SkyblockerConfig
                                 .controller(ConfigUtils::createBooleanController)
                                 .build())
                         .build())
+                        .option(Option.<Boolean>createBuilder()
+                                .name(Text.translatable("skyblocker.config.general.itemTooltip.showEssenceCost"))
+                                .binding(defaults.uiAndVisuals.showEssenceCost,
+                                        () -> config.uiAndVisuals.showEssenceCost,
+                                        newValue -> config.uiAndVisuals.showEssenceCost = newValue)
+                                .controller(ConfigUtils::createBooleanController)
+                                .build())
```
Comment (kevinthegreat1) [inline src/main/java/de/hysky/skyblocker/config/categories/GeneralCategory.java:231]: This config is a bit messed up. This should be in the `itemTooltip` group.

File: src/main/resources/assets/skyblocker/lang/en_us.json [hunk 5]
```diff
@@ -263,6 +263,7 @@
   "skyblocker.config.general.itemTooltip.enableMuseumInfo.@Tooltip": "If this item is donatable to the museum, then the item's category in the musuem is displayed. It also displays a marker indicating whether you've donated that item to your musuem or not (freebies not yet supported).\n\nMake sure to enable your Museum API for accurate information!",
   "skyblocker.config.general.itemTooltip.enableNPCPrice": "Enable NPC Price",
   "skyblocker.config.general.itemTooltip.enableObtainedDate": "Enable Obtained Date",
+  "skyblocker.config.general.itemTooltip.showEssenceCost": "Show Essence Cost",
```
Comment (kevinthegreat1) [inline src/main/resources/assets/skyblocker/lang/en_us.json:266]: Right now for all the price tooltips, it says "enable ...", but "show ..." is arguably better. So keep it as is?

**Review by kevinthegreat1 — CHANGES_REQUESTED** [review]

**Review by kevinthegreat1 — APPROVED** [review]

**Review by kevinthegreat1 — APPROVED** [review]

**Pull request merged**